quinn-proto/src/connection/mod.rs

```diff
@@ -2813,7 +2818,9 @@ where
 
     fn add_read_credits(&mut self, id: StreamId, transmit_max_stream_data: bool) {
         let space = &mut self.spaces[SpaceId::Data];
-        space.pending.max_data = true;
+        if self.streams.want_send_max_data() {
```

Comment:
What I'm not that happy with this approach is that it might be easy to miss the check before setting `pending`, and that not all paths which actually increase `local_max_data` might be covered. E.g. `stop` which is referenced higher up increased `max_data`, but did never set `pending`. And I think there is yet another path where we receive a `STREAM` frame for a stopped stream, increase `local_max_data` but don't set pending.

An alternative that I thought about is to encapsulate all the `self.local_max_data = self.local_max_data.saturating_add(buf.len() as u64);` into a `add_local_max_data(&mut self, increment: u64) -> WantTransmit` call, where `WantTransmit` is either a `bool` or a `#[must use]struct WantTransmit(bool)`, return that from every function that touches `local_max_data` and hope that the users then are reminded to touch `pending.max_data`.

But it's a bigger change - therefore I wanted to get some feedback first.

Reply:
Good catches re: MAX_DATA not being sent. I confirm the remaining issue you describe on stopped streams. Refactoring to lampshade that sounds good, though I might name the method `add_read_credits` to match `Connection`.

Reply:
lampshade? I will try the refactoring, and push if I like it better :)

Reply:
> lampshade?

Draw attention to.

Reply:
> An alternative that I thought about is to encapsulate all the self.local_max_data = self.local_max_data.saturating_add(buf.len() as u64); into a add_local_max_data(&mut self, increment: u64) -> WantTransmit call, where WantTransmit is either a bool or a #[must use]struct WantTransmit(bool), return that from every function that touches local_max_data and hope that the users then are reminded to touch pending.max_data.

This definitely sounds good to me!

Reply:
Will do. But also wouldn't mind get this version merged in between, since half of the changes would be the same - and the next change would then be pure refactoring without functional change.

Reply:
> without functional change

We do still need to fix the issue with no updates being sent when receiving frames on a stopped stream.

Reply:
> We do still need to fix the issue with no updates being sent when receiving frames on a stopped stream.

Has this been addressed?

Reply:
It now is added here too. It just initially wasn't, because it would have been addressed by the refactoring